rllib/agents/trainer.py

```diff
@@ -927,16 +982,105 @@ def step(self) -> ResultDict:
                 # Allow logs messages to propagate.
                 time.sleep(0.5)
                 raise e
+
+            # Stopping criteria.
+            if self.config["_disable_distributed_execution_api"]:
+                if self._by_agent_steps:
+                    sampled = self._counters[NUM_AGENT_STEPS_SAMPLED] - \
+                        agent_steps_sampled
+                    trained = self._counters[NUM_AGENT_STEPS_TRAINED] - \
+                        agent_steps_trained
+                else:
+                    sampled = self._counters[NUM_ENV_STEPS_SAMPLED] - \
+                        env_steps_sampled
+                    trained = self._counters[NUM_ENV_STEPS_TRAINED] - \
+                        env_steps_trained
+
+                min_t = self.config["min_time_s_per_reporting"]
+                min_sample_ts = self.config[
+                    "min_sample_timesteps_per_reporting"]
+                min_train_ts = self.config["min_train_timesteps_per_reporting"]
+                # Repeat if not enough time has passed or if not enough
+                # env|train timesteps have been processed (or these min values
+                # are not provided by the user).
+                if (not min_t or time.time() - time_start >= min_t) and \
+                        (not min_sample_ts or sampled >= min_sample_ts) and \
+                        (not min_train_ts or trained >= min_train_ts):
+                    break
             else:
                 break
 
-        # Still no result (even after n retries).
-        if result is None:
-            raise RuntimeError("Failed to recover from worker crash.")
-
         if hasattr(self, "workers") and isinstance(self.workers, WorkerSet):
+            # Sync filters on workers.
             self._sync_filters_if_needed(self.workers)
 
+            # Collect worker metrics.
+            if self.config["_disable_distributed_execution_api"]:
+                info = result.copy()
+                result["info"] = {LEARNER_INFO: info}
+
+                # Collect rollout worker metrics.
+                episodes, self._episodes_to_be_collected = collect_episodes(
```

Comment:
are we running stats on a new set of episodes?
this looks pretty weird that we collect a bunch of episodes after step_attemp() call, then just drop them?
btw, may be cleaner to put all these summarization code in a helper func.

Reply:
Yeah, `self._episodes_to_be_collected` are the ray object refs referencing those metrics results that timed out in the `ray.wait([w.get_metrics.remote() for w in self.workers])` call. Note that timeouts usually never happens as the default timeout threshold is 180sec.

Reply:
In other words, we never drop episodes. We always wait until they are completed before even sending them from the Workers to the driver and those that time out (`ray.wait()`) are collected in the next iteration.